nemo/src/io.rs

```diff
@@ -44,3 +45,52 @@ impl<W: Write> RecordWriter for csv::Writer<W> {
         Ok(())
     }
 }
+
+#[macro_export]
+/// Transparently read from a file that might be compressed.
+///
+/// # Example
+///
+/// ```ignore
+/// read_from_possibly_compressed_file!(file, compressed: |crdr| c, uncompressed: |urdr| u)
+/// ```
+/// reads from `file`, a [`PathBuf`][`std::path::PathBuf`]. If it is a
+/// gzip-compressed file, this calls `c` with `crdr` bound to the
+/// [`GzDecoder`][`flate2::read::GzDecoder`]. If it is not a
+/// gzip-compressed file, call `u` with `urdr` bound to the
+/// [`File`][`std::fs::File`].
+///
+/// ```ignore
+/// read_from_possibly_compressed_file!(file, |rdr| a)
+/// ```
+/// reads from `file`, a [`Path`][`std::path::PathBuf`]. This calls
+/// `a` with `rdr` bound to either the
+/// [`GzDecoder`][`flate2::read::GzDecoder`] or to the
+/// [`File`][`std::fs::File`].
+macro_rules! read_from_possibly_compressed_file {
```

Comment:
Any specific reason you are not using `Box<dyn Read>` (which would make this implementable as a function) ? Anyway I think we might want to clean this up later anyway, because the decompression should probably be factored out of the individual readers.

Reply:
For starters, `NTriplesReader` really needs a `BufRead`, not a `Read` (but we don't want a `BufRead` for DSV files, since the `csv` crate does it's own buffering internally). I agree, though, that this should, ideally, be handled before dispatching to the individual `TableReader`s.